File: feicai/backend/routers/scripts.py
from fastapi import APIRouter, HTTPException, Depends
import aiosqlite

from database import get_db
from schemas.script import (
    ScriptUpload,
    ScriptResponse,
    ScriptSplitRequest,
    ScriptSplitResponse,
    ScriptType,
)
from services.script_service import (
    save_script,
    read_script,
    split_script_by_ai,
    get_episode_info,
)

router = APIRouter(tags=["scripts"])


@router.post("/episodes/{episode_id}/script", response_model=ScriptResponse)
async def upload_script(episode_id: int, payload: ScriptUpload, db=Depends(get_db)):
    """上传剧本到指定集数"""
    episode = await get_episode_info(episode_id)
    if not episode:
        raise HTTPException(404, "集数不存在")

    try:
        file_path = await save_script(episode_id, payload.content, payload.script_type)
    except ValueError as e:
        raise HTTPException(400, str(e))

    return ScriptResponse(
        episode_id=episode_id,
        episode_number=episode["number"],
        has_script=True,
        content=payload.content[:500],  # 只返回前 500 字预览
        script_type=payload.script_type,
        file_path=file_path,
    )


@router.get("/episodes/{episode_id}/script", response_model=ScriptResponse)
async def get_script(episode_id: int):
    """获取指定集数的剧本"""
    episode = await get_episode_info(episode_id)
    if not episode:
        raise HTTPException(404, "集数不存在")

    script_data = await read_script(episode_id)
    if not script_data:
        return ScriptResponse(
            episode_id=episode_id,
            episode_number=episode["number"],
            has_script=False,
        )

    return ScriptResponse(
        episode_id=episode_id,
        episode_number=episode["number"],
        has_script=True,
        content=script_data["content"],
        script_type=script_data["script_type"],
        file_path=script_data["file_path"],
    )


@router.post("/episodes/{episode_id}/script/split", response_model=ScriptSplitResponse)
async def split_script(episode_id: int, payload: ScriptSplitRequest):
    """AI 分割全集剧本（传入内容）"""
    episode = await get_episode_info(episode_id)
    if not episode:
        raise HTTPException(404, "集数不存在")

    try:
        result = await split_script_by_ai(payload.content, payload.episode_count)
    except ValueError as e:
        raise HTTPException(400, str(e))

    return result


@router.delete("/episodes/{episode_id}/script")
async def delete_script(episode_id: int):
    """删除剧本"""
    episode = await get_episode_info(episode_id)
    if not episode:
        raise HTTPException(404, "集数不存在")

    script_data = await read_script(episode_id)
    if script_data and script_data["file_path"]:
        import os
        try:
            os.remove(script_data["file_path"])
        except OSError:
            pass

    return {"message": "剧本已删除"}

File: feicai/frontend/src/api/scripts.ts
const BASE = '/api'

export type ScriptType = 'traditional' | 'storyboard'

export interface ScriptUploadPayload {
  content: string
  script_type: ScriptType
  is_full_series?: boolean
}

export interface ScriptResponse {
  episode_id: number
  episode_number: number
  has_script: boolean
  content: string | null
  script_type: ScriptType | null
  file_path: string | null
}

export interface ScriptSplitPayload {
  content: string
  episode_count: number
}

export interface ScriptSplitResult {
  episode_number: number
  content: string
  start_marker: string | null
}

export interface ScriptSplitResponse {
  splits: ScriptSplitResult[]
  total_episodes: number
}

export async function uploadScript(
  episodeId: number,
  payload: ScriptUploadPayload
): Promise<ScriptResponse> {
  const res = await fetch(`${BASE}/episodes/${episodeId}/script`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })
  if (!res.ok) {
    const err = await res.json().catch(() => ({}))
    throw new Error(err.detail || '上传剧本失败')
  }
  return res.json()
}

export async function getScript(episodeId: number): Promise<ScriptResponse> {
  const res = await fetch(`${BASE}/episodes/${episodeId}/script`)
  if (!res.ok) throw new Error('获取剧本失败')
  return res.json()
}

export async function splitScript(
  episodeId: number,
  payload: ScriptSplitPayload
): Promise<ScriptSplitResponse> {
  const res = await fetch(`${BASE}/episodes/${episodeId}/script/split`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })
  if (!res.ok) {
    const err = await res.json().catch(() => ({}))
    throw new Error(err.detail || 'AI 分割失败')
  }
  return res.json()
}

